Implemented Relaxed Distributions

rachtsingh opened the pull request:
Not sure if this is the right name for this distribution (Concrete / GumbelSoftmax are other ideas), but this is what Tensorflow calls it. This PR uses the transforms machinery :)

I had to edit `TransformedDistribution`'s `log_prob` method to take into account `event_shape` - this is probably not the right way to do it, but a quick first try that makes it work. 

- [x] Implement RelaxedOneHotCategorical 
- [x]  Implement RelaxedBernoulli

cc @fritzo

rachtsingh committed:
Implemented RelaxedOneHotCategorical + RelaxedBernoulli distributions

fritzo commented on torch/distributions/relaxed_categorical.py:
```diff
@@ -0,0 +1,110 @@
+
+    Example::
+
+        >>> m = OneHotCategorical(torch.Tensor([ 0.25, 0.25, 0.25, 0.25 ]))
```
Could you update the docs and example sample?

fritzo commented on torch/distributions/relaxed_categorical.py:
```diff
@@ -0,0 +1,110 @@
+    return log_softmax(x, *args, **kwargs)
+
+
+class ExpRelaxedOneHotCategorical(Distribution):
```
`RelaxedCategorical` seems ok too, since `OneHot` is kind of the only relay to relax.

fritzo commented on torch/distributions/relaxed_categorical.py:
```diff
@@ -0,0 +1,110 @@
+
+
+def _log_softmax(x, *args, **kwargs):
+    if isinstance(x, torch.Tensor):
```
1. Since `Tensor` and `Variable` are being merged, this might not be necessary.
2. Prefer `torch.is_tensor()`, since `torch.Tensor` is an alias rather than a base class:
```py
isinstance(torch.FloatTensor([0]), torch.Tensor)
# True
isinstance(torch.DoubleTensor([0]), torch.Tensor)
# False
torch.Tensor is torch.FloatTensor
# True
```

rachtsingh replied:
1.  I'm hoping it's not necessary soon, but it definitely is right now because `log_softmax` takes `Variables` and raises when called on a `Tensor`. Can we leave the shim until the merge? 
2. Definitely, didn't realize this was the case.

fritzo commented on torch/distributions/relaxed_categorical.py:
```diff
@@ -0,0 +1,110 @@
+        [torch.FloatTensor of size 4]
+
+    Args:
+        probs (Tensor or Variable): event probabilities
```
Document `temperature` parameter

fritzo commented on torch/distributions/relaxed_categorical.py:
```diff
@@ -0,0 +1,110 @@
+    def rsample(self, sample_shape=torch.Size()):
+        sample_shape = torch.Size(sample_shape)
+        uniforms = clamp_probs(self.logits.new(self._extended_shape(sample_shape)).uniform_())
+        gumbels = uniforms.log().mul(-1).log().mul(-1)
```
Is `__neg__` cheaper or more stable than `.mul(-1)`?

rachtsingh replied:
Can I call it via `-(-(uniforms.log()).log())`?

fritzo replied:
Yes, I just assume it's a tiny bit cheaper.

fritzo commented on torch/distributions/relaxed_categorical.py:
```diff
@@ -0,0 +1,110 @@
+
+    def __init__(self, temperature, probs=None, logits=None):
+        super(RelaxedOneHotCategorical, self).__init__(ExpRelaxedOneHotCategorical(temperature, probs, logits),
+                                                       ExpTransform())
```
I would expect either the transform to be `ExpTransform().inv` or the base distribution to be `LogRelaxedOneHotCategorical`. Is the latter a reasonable renaming, analogous to `LogNormal`?

rachtsingh replied:
Right, I agree with `LogRelaxedOneHotCategorical` as well. However, TF uses the version I have because that's what the paper names it. My thinking is that it's probably more important just to document it and be consistent with the other interfaces, since no user will need to see it.

fritzo replied:
Sounds good, let's keep it standard.

fritzo commented on torch/distributions/transformed_distribution.py:
```diff
@@ -76,5 +76,7 @@ def log_prob(self, value):
             x = transform.inv(y)
             log_prob -= transform.log_abs_det_jacobian(x, y)
             y = x
+        if self.base_dist._event_shape:
```
Some transforms should already know their event shape. Also sometimes `log_prob` is just a number here (e.g. for `identity_transform`). I think it would be safer to do one of:
1. add an `event_shape` arg to pointwise transforms
2. implemement an `TensorTransform(base_transform, event_dim=1)` to wrap `ExpTransform`
3. make this hack a bit more robust:
```py
log_prob = ...sum of log_abs_det_jacobian_terms...
base_log_prob = self.base_dist.log_prob(y)  # knows the correct shape
if not isinstance(log_prob, numbers.Number):
    while log_prob.dim() > base_log_prob.dim():
        log_prob = log_prob.sum(-1)
log_prob += base_log_prob
```

rachtsingh replied:
I like solution `1.` the best. I think fixing the hack as you did doesn't work because `log_prob` isn't a number most of the time - for example it can be `[sample_shape x batch_shape x ...]`. Let me rethink this and get back to you.

fritzo replied:
:+1: We should also make `constraints` more aware of event dim. @tbrx is already doing this for multivariate normal in #52 and we'll need it e.g. for `AffineOperatorTransform` which transforms from `constraints.real_vector` to `constraints.real_vector`.

fritzo replied:
After thinking about this a bit, I think a 4th approach could be minimally invasive:
- add a static `event_dim` attribute to each `Transform` (only about 4 lines of code diff)
- update the `log_prob` accumulation loop in `TransformedDistribution.log_prob()`

I'll sketch this in a PR so just to discuss

fritzo reviewed (commented):
Nice factorization into 2 classes!

rachtsingh commented:
Ok, made changes based on comments, and added the `RelaxedBernoulli` distribution. Blocked on https://github.com/pytorch/pytorch/pull/4937 getting merged, which should be soon :)

sdmonov committed:
Move the cudnn include path before system include path (#5026)

In some cases when there are two different versions of cudnn installed,
one under /usr/local/cuda and other under a virtual env such as conda or
under the main system path /usr/include, the compiler would pickup the
cudnn.h from the virtual env/system path first. This is because cmake
generates C_INCLUDES and CXX_INCLUDES flags with system include path
first. All this may lead to linking problems as described in Issue #4869

Fixes #4869

rachtsingh commented:
Added tests that:

1. Rounding the RelaxedBernoulli distribution gives the corresponding Bernoulli distribution
2. Taking the argmax of the RelaxedOneHotCategorical gives the corresponding Categorical distribution
3. As the temperature becomes very large, the first consistently gives 0.5 and the latter gives equal values for each index.

zdevito committed:
Use blocks machinery to simplify bookkeeping in autodiff (#5036)

* Remove addValues and use WithInsertPoint

* Use blocks to simplify differentiate

Using @ezyang's suggestion, this change uses a block rather than
staging annotations to represent the reverse pass. This allows us
to reuse the machinery to copy graphs/blocks to extract the
reverse pass concisely.

This also change the input order of Gradients df to:
   [output vjps][temporary vjps][captures]

In addition to being simpler to generate in this order, it also
will allow ExecutionPlan to append the captures onto the already-
existing input list of vjps that are given by the autograd,
rather than have to prepend them, which should be slightly cheaper.

* Enforce that input capture are before outputs

This changes the Gradient struct to enforce that input
captures appear before output captures in the capture list,
which makes it easier to use in ExecutionPlan.

gchanan committed:
Don't allow scalars where vectors are required in mv, addmv, ger, addr. (#5003)

* Don't allow scalars where vectors are required in mv, addmv, ger, addr.

* Fix scalar_tensor_test for ger.

* Address review comments.

* Fix merge.

colesbury committed:
Operate on Variables in torch.nn.init (#4964)

Once Variable and Tensor are merged the existing Variable test would
cause an infinite recursion. Instead, modify the Variables directly
inside a `no_grad()` block.

zou3519 committed:
Expose sparse variable addmm, addmm_ (#5016)

sspaddmm, mm for sparse tensors to come in another pr; they're a little more involved.

fritzo reviewed (approved):
LGTM

Commit:
add reduce=True argument to MultiLabelMarginLoss (#4924)

* add reduce=True argument to MultiLabelMarginLoss

* Fix lint

* Addressed comments

* Remove unneeded syncthreads calls

rachtsingh commented on torch/distributions/relaxed_categorical.py:
```diff
@@ -0,0 +1,115 @@
+    def rsample(self, sample_shape=torch.Size()):
+        sample_shape = torch.Size(sample_shape)
+        uniforms = clamp_probs(self.logits.new(self._extended_shape(sample_shape)).uniform_())
+        gumbels = -((-(uniforms.log())).log())
```
We can definitely implement this via a `TransformedDistribution`  of `Gumbel + BoltzmannTransform`; I just didn't see the `BoltzmannTransform`. I'll make an issue about implementing the `log_abs_det_jacobian` and then revisit it when that's been solved?

fritzo replied:
I think this will be tricky, but let's move discussion to that issue. Feel free to send this PR upstream before the refactoring.

royf commented:
@rachtsingh @fritzo 
Are there plans to add KL divergence between two RelaxedOneHotCategorical distributions?
It should be possible to implement it based on this work: https://mast.queensu.ca/~communications/Papers/gil-msc11.pdf (see Proposition 2.3.23).
I may be able to contribute if you could please share a pointer to the math behind the current `log_prob` implementation in https://github.com/pytorch/pytorch/blob/355b24438c1d48ab98948c2ad0c426c6b19f029d/torch/distributions/relaxed_categorical.py#L75.
